shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -634,125 +639,239 @@ static bool ClipRRectContainsPlatformViewBoundingRect(const SkRRect& clip_rrect,
 }
 
 DlCanvas* FlutterPlatformViewsController::CompositeEmbeddedView(int64_t view_id) {
-  // Any UIKit related code has to run on main thread.
-  FML_DCHECK([[NSThread currentThread] isMainThread]);
-  // Do nothing if the view doesn't need to be composited.
-  if (views_to_recomposite_.count(view_id) == 0) {
-    return slices_[view_id]->canvas();
-  }
-  CompositeWithParams(view_id, current_composition_params_[view_id]);
-  views_to_recomposite_.erase(view_id);
   return slices_[view_id]->canvas();
 }
 
 void FlutterPlatformViewsController::Reset() {
-  for (int64_t view_id : active_composition_order_) {
-    UIView* sub_view = root_views_[view_id].get();
-    [sub_view removeFromSuperview];
-  }
-  root_views_.clear();
-  touch_interceptors_.clear();
-  views_.clear();
+  fml::TaskRunner::RunNowOrPostTask(
+      platform_task_runner_, [&, composition_order = composition_order_]() {
+        for (int64_t view_id : composition_order_) {
+          [platform_views_[view_id].root_view.get() removeFromSuperview];
+        }
+        platform_views_.clear();
+      });
+
   composition_order_.clear();
-  active_composition_order_.clear();
   slices_.clear();
   current_composition_params_.clear();
-  clip_count_.clear();
   views_to_recomposite_.clear();
   layer_pool_->RecycleLayers();
   visited_platform_views_.clear();
 }
 
-SkRect FlutterPlatformViewsController::GetPlatformViewRect(int64_t view_id) {
-  UIView* platform_view = GetPlatformViewByID(view_id);
-  UIScreen* screen = [UIScreen mainScreen];
-  CGRect platform_view_cgrect = [platform_view convertRect:platform_view.bounds
-                                                    toView:flutter_view_];
-  return SkRect::MakeXYWH(platform_view_cgrect.origin.x * screen.scale,    //
-                          platform_view_cgrect.origin.y * screen.scale,    //
-                          platform_view_cgrect.size.width * screen.scale,  //
-                          platform_view_cgrect.size.height * screen.scale  //
-  );
-}
-
 bool FlutterPlatformViewsController::SubmitFrame(GrDirectContext* gr_context,
                                                  const std::shared_ptr<IOSContext>& ios_context,
-                                                 std::unique_ptr<SurfaceFrame> frame) {
+                                                 std::unique_ptr<SurfaceFrame> background_frame) {
   TRACE_EVENT0("flutter", "FlutterPlatformViewsController::SubmitFrame");
 
-  // Any UIKit related code has to run on main thread.
-  FML_DCHECK([[NSThread currentThread] isMainThread]);
-  if (flutter_view_ == nullptr) {
-    return frame->Submit();
+  if (flutter_view_ == nullptr || (composition_order_.empty() && !had_platform_views_)) {
```

Comment:
Consider adding a comment here to eliminate the minor brain power required to work out that empty `composition_order_` means no platform views, means no work to do:
```suggestion
  // No platform views to render; we're done.
  if (flutter_view_ == nullptr || (composition_order_.empty() && !had_platform_views_)) {
```

Reply:
Done